docs/adr/0006-config-driven-content-types.md

```diff
@@ -0,0 +1,125 @@
+
+#### 4. Backwards compatibility
+
+We need to be sending exactly the same payload to Publishing API (and in turn, Content Store and the frontend rendering apps) as we have been doing in the lead up to config-driven document types. This is to enable us to work at our own pace and gain the value of the refactor more quickly.
+
+We do, however, need to be able to easily change the payload in response to changing business requirements such as flexible pages and multi-platform support. Without prescribing the exact detail of such a system, this ADR suggests that config-driven content types could have a shared generic presenter, called at the point of publishing a document’s payload to Publishing API.
+
+In that hypothetical system, by default, we would be sending a genericised or ‘flexible’ payload to Publishing API, and the applications downstream of Publishing API may not know how to render such a payload. We could then build a ‘legacy presenter’ to take a generic/flexible payload and map it to the existing (‘legacy’) payload structure, so that as far as Publishing API, Content Store and Frontend are concerned, nothing has changed. We could maintain this low-cost backwards compatibility solution for as long as necessary, only omitting the call to the legacy presenter when ready.
+
+#### 5. Forwards compatibility
+
+When schema changes are required to the content types, we need to make sure that we can release those migrations safely without breaking existing content.
+
+It is outside of the scope of this ADR to propose a full plan for how these schema migrations will be managed, but generally we need to be mindful of the fact that removing an attribute from the schema, or changing its “key”, will make any existing content which has data stored for any such attributes incompatible with the schema. Similarly, adding a new attribute with the “required” property would make existing content invalid.
+
+We therefore need to apply the same rigour that we would with normal database migrations to our schema migrations, but we won’t have MySQL’s built-in integrity checks to help us. We should consider building some tooling to support these migrations, such as:
+
+1. a tool that parses all of the schemas during CI and prevents/warns about dangerous changes
+2. a bulk update tool that can help us set or change data for an attribute across multiple content items during a migration
+
+## Consequences
+
+## Benefits
+
+- Clarity: behaviour is declared in one place per content type.
+- Consistency: reduces accidental divergence between similar types.
+- Maintainability: easier to add, merge, or retire types.
+- Composability: new types can easily be created from a selection of existing behaviours
+- Scalability: enables more flexibility in Whitehall’s output, to accommodate multi platform requirements, personalisation, and publisher-led layouts.
```

Comment:
Publisher-led layouts now in theory supported in Frontend: https://github.com/alphagov/frontend/pull/4968/commits/8ea7bf8d22d61aeadfb7e0fce9773198658d0ca7